### PR TITLE
Add tests for casadm --set-param and --get-param options check

### PR DESCRIPTION
**Pull request by rafalste**
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@intel.com>

### CODE REVIEW

File: test/functional/tests/cli/test_set_get_params.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+            # Check default parameters for other cores:
+            for j in range(1, cores_per_cache):
+                check_seqcutoff_parameters(core[0][j], default_seqcutoff_params)
+            for i in range(1, caches_count):
+                for j in range(cores_per_cache):
+                    check_seqcutoff_parameters(core[i][j], default_seqcutoff_params)
+
+    with TestRun.step(
+        "Set new random values for sequental cut-off parameters "
```
Comment (Deixx) [inline test/functional/tests/cli/test_set_get_params.py]: sequential

File: test/functional/tests/cli/test_set_get_params.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+            for j in range(cores_per_cache):
+                check_seqcutoff_parameters(core[0][j], random_seqcutoff_params)
+
+            # Check default parameters for other cache instances:
+            for i in range(1, caches_count):
+                for j in range(cores_per_cache):
+                    check_seqcutoff_parameters(core[i][j], default_seqcutoff_params)
+
+    with TestRun.step(
+        "Set new random values for sequental cut-off parameters for all cores"
```
Comment (Deixx) [inline test/functional/tests/cli/test_set_get_params.py]: sequential

File: test/functional/tests/cli/test_set_get_params.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        cache, core = cache_prepare(cache_mode, cache_dev, core_dev)
+
+    with TestRun.step("Check sequential cut-off default parameters"):
+        default_seqcutoff_params = SeqCutOffParameters.default_seq_cut_off_params()
+        for i in range(caches_count):
+            for j in range(cores_per_cache):
+                check_seqcutoff_parameters(core[i][j], default_seqcutoff_params)
+
+    with TestRun.step(
+        "Set new random values for sequental cut-off parameters for one core only"
```
Comment (Deixx) [inline test/functional/tests/cli/test_set_get_params.py]: sequential

File: test/functional/tests/cli/test_set_get_params.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+          - All sequential cut-off parameters are set to given values.
+          - All sequential cut-off parameters displays proper values.
+    """
+
+    with TestRun.step("Partition cache and core devices"):
+        cache_dev, core_dev = storage_prepare()
+
+    with TestRun.step(
+        f"Start {caches_count} cache(s) in {cache_mode} cache mode "
+        f"and add {cores_per_cache} core(s) per cache"
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    cache_dev.create_partitions(cache_parts)
+    core_dev = TestRun.disks["core"]
+    core_parts = [Size(2, Unit.GibiByte)] * cores_per_cache * caches_count
+    core_dev.create_partitions(core_parts)
+
+    return cache_dev, core_dev
+
+
+def cache_prepare(cache_mode, cache_dev, core_dev):
+    if caches_count < 2 or cores_per_cache < 2:
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        TestRun.fail(
+            "There should be at least 2 cache instances and 2 cores per cache "
+            "for this test to run correctly."
+        )
+    cache = []
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    for i in range(caches_count):
+        cache.append(
+            casadm.start_cache(cache_dev.partitions[i], cache_mode, force=True)
+        )
+    core = [[0 for j in range(cores_per_cache)] for i in range(caches_count)]
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    """
+
+    with TestRun.step("Partition cache and core devices"):
+        cache_dev, core_dev = storage_prepare()
+
+    with TestRun.step(
+        f"Start {caches_count} cache(s) in {cache_mode} cache mode "
+        f"and add {cores_per_cache} core(s) per cache"
+    ):
+        cache, core = cache_prepare(cache_mode, cache_dev, core_dev)
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+
+            # Check default parameters for other cache instances:
+            for i in range(1, caches_count):
+                for j in range(cores_per_cache):
+                    check_seqcutoff_parameters(core[i][j], default_seqcutoff_params)
+
+    with TestRun.step(
+        "Set new random values for sequental cut-off parameters for all cores"
+    ):
+        for check in range(number_of_checks):
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    """
+
+    with TestRun.step("Partition cache and core devices"):
+        cache_dev, core_dev = storage_prepare()
+
+    with TestRun.step(
+        f"Start {caches_count} cache(s) in {cache_mode} cache mode "
+        f"and add {cores_per_cache} core(s) per cache"
+    ):
+        cache, core = cache_prepare(cache_mode, cache_dev, core_dev)
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        "Set new random values for sequental cut-off parameters for all cores"
+    ):
+        for check in range(number_of_checks):
+            for i in range(caches_count):
+                for j in range(cores_per_cache):
+                    random_seqcutoff_params = new_seqcutoff_parameters_random_values()
+                    core[i][j].set_seq_cutoff_parameters(random_seqcutoff_params)
+                    check_seqcutoff_parameters(core[i][j], random_seqcutoff_params)
+
+    with TestRun.step("Stop all caches"):
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+            for i in range(1, caches_count):
+                check_cleaning_parameters(
+                    cache[i], cleaning_policy, default_cleaning_params
+                )
+
+    with TestRun.step(
+        f"Set new random values for {cleaning_policy} "
+        f"cleaning policy parameters for all cache instances"
+    ):
+        for check in range(number_of_checks):
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+                )
+                if cleaning_policy == CleaningPolicy.alru:
+                    cache[i].set_params_alru(random_cleaning_params)
+                if cleaning_policy == CleaningPolicy.acp:
+                    cache[i].set_params_acp(random_cleaning_params)
+                check_cleaning_parameters(
+                    cache[i], cleaning_policy, random_cleaning_params
+                )
+
+    with TestRun.step("Stop all caches"):
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    for i in range(caches_count):
+        for j in range(cores_per_cache):
+            core_partition_nr = i * cores_per_cache + j
+            core[i][j] = cache[i].add_core(core_dev.partitions[core_partition_nr])
+
+    return cache, core
+
+
+def new_seqcutoff_parameters_random_values():
+    threshold_random_value = Size(random.randrange(1, 1000000), Unit.KibiByte)
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    policy_random_value = random.choice(
+        [SeqCutOffPolicy.always, SeqCutOffPolicy.full, SeqCutOffPolicy.never]
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+
+    return seqcutoff_params
+
+
+def new_cleaning_parameters_random_values(cleaning_policy):
+    if cleaning_policy == CleaningPolicy.alru:
+        wake_up_time_random_value = Time(seconds=random.randrange(1, 3600))
+        staleness_time_random_value = Time(seconds=random.randrange(1, 3600))
+        flush_max_buffers_random_value = random.randrange(1, 10000)
+        activity_threshold_random_value = Time(milliseconds=random.randrange(500, 1000000))
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+        cleaning_params = FlushParametersAlru()
+        cleaning_params.wake_up_time = wake_up_time_random_value
+        cleaning_params.staleness_time = staleness_time_random_value
+        cleaning_params.flush_max_buffers = flush_max_buffers_random_value
+        cleaning_params.activity_threshold = activity_threshold_random_value
+
+    if cleaning_policy == CleaningPolicy.acp:
+        wake_up_time_random_value = Time(milliseconds=random.randrange(0, 10000))
+        flush_max_buffers_random_value = random.randrange(1, 10000)
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py (hunk 1 above)
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py (hunk 2 above)
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

File: test/functional/tests/cli/test_set_get_params.py (hunk 3 above)
Comment (rafalste) [reply test/functional/tests/cli/test_set_get_params.py]: Done

**Comment by rafalste** [comment]
retest this please

File: test/functional/api/cas/casadm.py [hunk 18]
```diff
@@ -288,7 +288,7 @@ def set_param_cleaning_acp(cache_id: int, wake_up: int = None, flush_max_buffers
     output = TestRun.executor.run(
         set_param_cleaning_acp_cmd(
             cache_id=str(cache_id),
-            wake_up=str(wake_up) if wake_up else None,
+            wake_up=str(wake_up) if wake_up is not None else None,
```
Comment (Ostrokrzew) [inline test/functional/api/cas/casadm.py]: Is it necessary? You didn't change similar line below and I think it works.

Comment (rafalste) [reply test/functional/api/cas/casadm.py]: 'wake_up_time' in ACP ranges from 0 not 1 like in ALRU, so when it is set to 0 this 'if' statement treats it like False and don't pass that 0 value to 'set_param_cleaning_acp_cmd' function. That change I've made strictly compare wake_up to None, so when it's 0 then everything works as it should.

Comment (Ostrokrzew) [reply test/functional/api/cas/casadm.py]: Thanks for explanation.

**Pull request merged**